### ISSUE TITLE
Excel: Content Add-ins Leak Memory Eventually Crashing Add-in

### ISSUE THREAD

**Issue by superstructor** [OfficeDev/office-js#1054]
> Message from office-js bot:
We’re closing this issue because it has been inactive for a long time. We’re doing this to keep the issues list manageable and useful for everyone. If this issue is still relevant for you, please create a new issue.
> Thank you for your understanding and continued feedback.

# Content Add-ins Leak Memory Eventually Crashing Add-in

## Expected Behavior

**Content Add-ins** do not leak memory.

## Current Behavior

**Content Add-ins** leak memory on every call to `Excel.run()` at a rate
large enough to eventually crash the add-in.

**TaskPane Add-ins** with exactly the same code do not display the same behaviour.

The amount of memory leaked ranges from 10MB to 100MB per `Excel.run()` call.

The amount of memory leaked correlates with the complexity of the `Excel.run()` 
batch function, such as the number of proxy objects and `Excel.sync()`
calls used. 

Even an empty batch function leaks memory at a rate high enough to consume host
memory and crash the Add-in. 

Even code this trivial leaks up to 10MB per call:
```js
export async function run() {
  try {
    await Excel.run(async context => {
      return Promise.resolve(true);
    });
  } catch (error) {
    console.error(error);
  }
}
```

The memory leak is not permanent. After about twenty minutes it is freed but by 
then the Add-in may have crashed already due to memory pressure, if many calls
to `Excel.run()` have been made.

An almost trivial project which reproduces the problem is [available here](https://github.com/day8/office-js-memory-leak). It is based on the Microsoft Yeoman office-js generator.

The following gif shows what happens when I run the project and repeatedly click the 'Run' button.
The memory usage of Excel in Task Manager constantly increases, to the point that many gigabytes will
be consumed. Mysteriously, twenty minutes in the future the memory is reclaimed without any user interaction
but not before this accumulated use of memory causes system or Add-in instablity problems.

![office-js 2020-03-20 13-44](https://user-images.githubusercontent.com/350450/77129910-a309a100-6aba-11ea-91b5-99abb5d4276f.gif)

This same project run as a **TaskPane Add-in** does not cause memory usage to increase at all.

## Steps to Reproduce, or Live Example

An almost trivial project which reproduces the problem is [available here](https://github.com/day8/office-js-memory-leak). It is based on the Microsoft Yeoman office-js generator.

## Context

Build a Content Add-in for Excel. 

This problem makes production deployment an issue as the add-in can crash due to memory pressure which inconveniences users.

## Your Environment
* Platform : PC
* Host: Excel
* Office version number: 16.0.12527.20260 64-bit
* Operating System: Window 10 1909

**Comment by keyur32** [comment]
Hey @superstructor, thanks for the awesome repro steps. Unfortunately, I have not been able to repro this on my end. I tried the sample and the memory footprint stays the same.  

Question: Are you using IE or the Edge Webview? From the information provided it looks like the latter but I'm not sure. Can you expand the Excel.exe instance in task manager and share that screenshot please?

**Comment by superstructor** [comment]
![Screenshot from 2020-03-23 11-20-40](https://user-images.githubusercontent.com/350450/77262104-5a3f2b80-6cf8-11ea-93f5-9cd036be9e2f.png)

**Comment by superstructor** [comment]
It'll be Edge Webview as I can connect to it with the standalone Edge Developer Tools @keyur32 

Are you sure your opening it as a **Content** Add-in ? The button in the top ribbon does not have a open **Content** Add-in command so you need to open as per the reproduction steps I just added to the README of the project to be clearer:

[https://github.com/day8/office-js-memory-leak/blob/master/README.md#setup](https://github.com/day8/office-js-memory-leak/blob/master/README.md#setup)

If its not clear feel free to reach out with questions or I am even happy to connect via Google Meet if you want me to guide you through the reproduction.

**Comment by superstructor** [comment]
For clarity ... I'm going to restate the issue ... 

1. The issue we are reporting is a kind of "memory leak"
2. It can result in Excel memory usage growing effectively unbounded ... initially you will see it rise by at least 10s of MBs adding up to 100s of MB and certainly we've easily made it go to 10s of GBs of memory if you call `Excel.run()` enough times,
3. The memory which is "leaked" eventually gets reclaimed.  But only after something like 20 mins. 
4. But, by that time, Excel could have completely blown up the user's machine through excessive memory use 

**Causing It**

1. create a `content pane`  (not a `task pane`!!!) 
2. call  `Excel.run` from within that `content pane`
3. Observe Excel memory usage increase
4. The amount of memory "leaked" increases with the amount of "work" done within the batch function given to  `Excel.run`. 

Yeah, it is that simple.  

Just to be clear ... this memory leak does NOT occur in a `task pane` (we've not tested in a dialog pane). 

**Replicating It**

1.  we have created a minimal reproducing repository https://github.com/day8/office-js-memory-leak/blob/master/README.md
2. We used Microsoft's Yeoman office-js generator to create the code in the repository
3. We then made the most minimal set of changes possible to the template 
4. The batch function we pass to `Excel.run` is as minimal as possible
2. It has a clear step-by-step instructions which you should follow

**Comment by lindalu-MSFT** [comment]
@lumine2008 Hi Raymond, can you shed any light on this issue?

**Comment by mike-thompson-day8** [comment]
@lindalu-MSFT  can you explain the resolution of this, please?  The reason for closing? 

We carefully provided a repo which replicated the problem. ***We then participated in a screenshare session with members of the Microsoft team and they acknowledged the problem is real.***   This is clearly a serious problem.

**Comment by lindalu-MSFT** [comment]
@mike-thompson-day8 Misunderstanding! This was tagged resolved. I removed the resolved tag and readded the needs attention. In addition, I added under investigation so it won't get closed again. The issue is reopened!

**Comment by lindalu-MSFT** [comment]
@lliu113 Hi, reassigning to you as this appears to be a content add-in issue. Can you help with this? Thanks in advance!

**Comment by lliu113** [comment]
Reassigning to @croicu who owns content add-ins on desktop

**Comment by chanhohang** [comment]
[Updated on 15 July 2021]
I just realize this issue is reporting memory issue in Content Add-ins while my problem is in WebView2 Task pane add in, i create a separate issue to track my issue https://github.com/OfficeDev/office-js/issues/1988

______
Hi,

My Excel Add-in is also hitting into an Out of Memory Problem when running in WebView2 Runtime.

I did a memory snapshot in the middle of my testing and here is the snapshot to show the heap memory object.
![image](https://user-images.githubusercontent.com/4111001/123902969-28b01800-d9b1-11eb-9d73-e9566e6e0c24.png)


agaveResponseCallback is of concern in particular from the heap dump.

When i google this, i can see a similar issue is found in Outlook add-in https://github.com/OfficeDev/office-js/issues/584

My Add-in is running in Office365 and my environment is below.

- Browser info = {"BrowserName":"Chromium based edge (dev or canary)","FullVersion":"91.0.864.59","MajorVersion":91,"AppName":"Netscape","UserAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.114 Safari/537.36 Edg/91.0.864.59"}
- Excel API version = 1.13
- Office Excel version ![image](https://user-images.githubusercontent.com/4111001/123903261-b7249980-d9b1-11eb-8ba1-987b1a3378ff.png)

When i expand the heap dump and i can see a lot of agaveResponseCallback which is holding a json strong similar to below.
`{"ActionId":10356,"Value":{"_ReferenceId":"R!{B0EA95B6-6B87-47F2-A83F-4273C1AAD5EE}!{969774EB-5BBA-4394-859B-3FDB4DCC66C7}"}},{"ActionId":10357,"Value":null},{"ActionId":10359,"Value":{"_ReferenceId":"R!{7C52A738-365E-4A08-85BA-2DB3CCC6FEF4}!{969774EB-5BBA-4394-859B-3FDB4DCC66C7}"}},{"ActionId":10360,"Value":null},{"ActionId":10361,"Value":{"_ReferenceId":"R!{7C52A738-365E-4A08-85BA-2DB3CCC6FEF4}!{969774EB-5BBA-4394-859B-3FDB4DCC66C7}"}},{"ActionId":10362,"Value":null},{"ActionId":10364,"Value":{"_ReferenceId":"R!{E9A2033D-F961-480F-AD96-0EDDD8EA959C}!{969774EB-5BBA-4394-859B-3FDB4DCC66C7}"}},{"ActionId":10365,"Value":null},{"ActionId":10366,"Value":{"_ReferenceId":"R!{E9A2033D-F961-480F-AD96-0EDDD8EA959C}!{969774EB-5BBA-4394-859B-3FDB4DCC66C7}"}},{"ActionId":10367,"Value":null},{"ActionId":10369,"Value":{"_ReferenceId":"R!{6CE39F83-C138-409B-9120-5943E48DD6E5}!{969774EB-5BBA-4394-859B-3FDB4DCC66C7}"}},{"ActionId":10370,"Value":null},{"ActionId":10371,"Value":{"_ReferenceId":"R!{6CE39F83-C138-409B-9120-5943E4"@36424961}`


May I know the status of this memory leak and if this is addressed in the coming release?
